executor/stale_txn_test.go

```diff
@@ -772,3 +772,49 @@ func (s *testStaleTxnSuite) TestSetTransactionInfoSchema(c *C) {
 	tk.MustExec("commit")
 	c.Assert(tk.Se.GetInfoSchema().SchemaMetaVersion(), Equals, schemaVer3)
 }
+
+func (s *testStaleTxnSuite) TestStaleSelect(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	defer tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (id int)")
+	tk.MustExec("insert into t values (1)")
+
+	time.Sleep(10 * time.Millisecond)
+	time1 := time.Now()
+	tk.MustExec("insert into t values (2)")
+	time.Sleep(10 * time.Millisecond)
+	time2 := time.Now()
+	tk.MustExec("insert into t values (3)")
+
+	staleRows := testkit.Rows("1")
+	staleSQL := fmt.Sprintf(`select * from t as of timestamp '%s'`, time1.Format("2006-1-2 15:04:05.000"))
```

Comment:
What will happen in this situation?
```
create table t (id int)
insert into t values (1)                                         ---- 1s
execute select * from t as of timestamp Now() - 2 Second;        ----- 4s
alter table t add column age int;                                ----- 5s
execute select * from t as of timestamp Now() - 2 Second;        ----- 8s
```

Reply:
I wanted to prepare the handle of infoschema in #25080, but that is not true anymore. The handle of infoschema is now added.

Reply:
I want to get the values ''(2, 2) "in the time point of  `8s` . but I guess you implement will always get the result `1`.

Reply:
Yeah, the timestamp is fixed once preprocessed now.

Your expectation needs a re-preprocessing, which drops all the prepared efforts. It is not a simple reason of my implementation. If timestamp is changed, schema and snapshot is changed, the result from previous prepare statements is meaningless.

Reply:
I could provide a hack solution, if you really want it to be true.

Reply:
Yes, I think it's the main user scenario, people want to read recent data rather than fixed in a point. and it's only re-preprocessing on the `stale read`, I think it should be accepted.

Reply:
No, it must be re-preprocessed. Infoschema may change. Even if it is not changed, a check for stale infoschema is needed, which leads to one more tikv metadata access.

> Agree with you.

Glad to hear that but it can not be true for now. I can only hack around. The solution will be tricky.

Reply:
Even if we have the infoschema cache, still  we need to go to access TiKV to get the schema version for every request?

Reply:
No for the current solution. Because it is already ensured when `prepare`. If you want to ensure the consistency for `NOW() - 2 seconds`, however, you don't know the schema version before 2 seconds. It is not `prepared` due to `NOW()`.

FYI, check https://github.com/pingcap/tidb/blob/5c95062cc34d6d37e2e921f9bddba6205b43ee3a/domain/domain.go#L103-L107 for the code of getting `SchemaVersion`.

Reply:
I have checked it, the implementation is different from what I understand. I think the schema version is guaranteed to increase, it could be optimized to reduce the extra request by check the schema version is not changed between the older request and the up-to-date schema version.

Reply:
Otherwise, it is easy to have a single point of bottleneck, because the meta information is always read from a certain TiKV node.